bqskit/runtime/manager.py

```diff
@@ -0,0 +1,387 @@
+
+                    if msg == RuntimeMessage.SUBMIT:
+                        task = cast(RuntimeTask, payload)
+                        self._assign_new_task(task)
+
+                    elif msg == RuntimeMessage.SUBMIT_BATCH:
+                        tasks = cast(List[RuntimeTask], payload)
+                        self._assign_new_tasks(tasks)
+
+                    elif msg == RuntimeMessage.RESULT:
+                        result = cast(RuntimeResult, payload)
+                        self._handle_result_coming_down(result)
+
+                    elif msg == RuntimeMessage.CANCEL:
+                        addr = cast(RuntimeAddress, payload)
+                        self._handle_cancel(addr)
+
+                    elif msg == RuntimeMessage.SHUTDOWN:
+                        self._handle_shutdown()
+                        return
+
+                else:
+
+                    if msg == RuntimeMessage.SUBMIT:
+                        task = cast(RuntimeTask, payload)
+                        self._recieve_new_task(task)
+
+                    elif msg == RuntimeMessage.SUBMIT_BATCH:
+                        tasks = cast(List[RuntimeTask], payload)
+                        self._recieve_new_tasks(tasks)
+
+                    elif msg == RuntimeMessage.RESULT:
+                        result = cast(RuntimeResult, payload)
+                        self._handle_result_going_up(result)
+
+                    else:
+                        # Forward all other messages up
+                        self.outgoing.append((self.upstream, msg, payload))
+
+    def _handle_shutdown(self) -> None:
+        """Shutdown the manager and clean up spawned processes."""
+        if not self.running:
+            return
+
+        # Stop running
+        self.logger.info('Shutting down server.')
+        self.running = False
+
+        # Instruct workers to shutdown
+        if self.workers is not None:
+            for wconn, _ in self.workers:
+                try:
+                    wconn.send((RuntimeMessage.SHUTDOWN, None))
+                    wconn.close()
+                except Exception:
+                    pass
+
+            for _, p in self.workers:
+                if p is not None:
+                    p.join()
+                    self.logger.debug('Joined worker.')
+
+            self.workers.clear()
+            self.workers = None  # type: ignore
+
+        # Join threads
+        if self.outgoing_thread is not None:
+            self.outgoing_thread.join()
+            self.outgoing_thread = None  # type: ignore
+            self.logger.debug('Joined outgoing thread.')
+
+        # Forward shutdown message upwards if possible
+        try:
+            self.upstream.send((RuntimeMessage.SHUTDOWN, None))
+            self.upstream.close()
+        except Exception:
+            pass
+
+    def _recieve_new_task(self, task: RuntimeTask) -> None:
+        """Either send the task upstream or schedule it downstream."""
```

Comment:
BIG ISSUE - This will create "waiting" threads and consume more and more from the system, because if a worker sent a job and he is waiting for it, then its thread can't be used, and this can have a catastrophic effect.

The runtime needs to know if the thread that send the new task is waiting for completion, so it could use its resources

Reply:
The worker can be used while it has a waiting task. The manager/server will not schedule new tasks on a worker with 1 waiting tasks. That is, until all workers have 1 task. Tasks are just sent to the worker with the fewest number of known tasks. I think the reason for this confusion may have been the naming of `idle_resources` fields rather than something like `num_tasks`. Thoughts?